pkg/sentry/fsimpl/erofs/directory.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2023 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package erofs
+
+import (
+	"sort"
+	"sync"
+
+	"gvisor.dev/gvisor/pkg/abi/linux"
+	"gvisor.dev/gvisor/pkg/context"
+	"gvisor.dev/gvisor/pkg/errors/linuxerr"
+	"gvisor.dev/gvisor/pkg/sentry/vfs"
+)
+
+func (i *inode) getDirents() ([]vfs.Dirent, error) {
+	// Fast path.
+	i.dirMu.RLock()
+	dirents := i.dirents
+	i.dirMu.RUnlock()
+	if dirents != nil {
+		return dirents, nil
+	}
+
+	// Slow path.
+	i.dirMu.Lock()
+	defer i.dirMu.Unlock()
+
+	off := int64(1)
+	if err := i.IterDirents(func(name string, typ uint8, nid uint64) error {
+		dirents = append(dirents, vfs.Dirent{
+			Name:    name,
+			Type:    linux.FileTypeToDirentType(typ),
+			Ino:     nid,
+			NextOff: off,
+		})
+		off++
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	// "." and ".." should always be presented.
+	if len(dirents) < 2 {
+		return nil, linuxerr.EUCLEAN
+	}
+
+	i.dirents = dirents
+	return dirents, nil
+}
+
+func (i *inode) lookup(name string) (uint64, error) {
+	// TODO: We need optimize this. For simplicity, currently a lookup will
+	// cause all dirents to be read and cached. But it hurts the performance
+	// of large directories.
+	dirents, err := i.getDirents()
+	if err != nil {
+		return 0, err
+	}
+
+	// Skip "." and ".."
+	dirents = dirents[2:]
+
+	// The dirents are sorted in alphabetical order. We do a binary search
+	// to find the target.
+	idx := sort.Search(len(dirents), func(i int) bool {
+		return dirents[i].Name >= name
+	})
+	if idx >= len(dirents) || dirents[idx].Name != name {
+		return 0, linuxerr.ENOENT
+	}
+	return dirents[idx].Ino, nil
+}
+
+func (d *dentry) lookup(ctx context.Context, name string) (*dentry, error) {
+	// Fast path, dentry already exists.
+	d.dirMu.RLock()
+	child, ok := d.childMap[name]
+	d.dirMu.RUnlock()
+	if ok {
+		return child, nil
+	}
+
+	// Slow path, create a new dentry.
+	d.dirMu.Lock()
+	defer d.dirMu.Unlock()
+	if child, ok := d.childMap[name]; ok {
+		return child, nil
+	}
+
+	nid, err := d.inode.lookup(name)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Another optimization idea: Implement negative dentry entries.

- Instead of returning ENOENT from `inode.lookup()`, make it return `0, nil` when an inode is not present.
- In `dentry.lookup()`, when we see that `0, nil` is returned, we should cache a negative dentry entry in `childMap`: `childMap[name] = nil`. So future queries for this name should return immediately without having to query the disk.
- `dentry.lookup()` can continue returning ENOENT, so callers don't need to be updated.

We can do this in a follow-up change too. In practice we have found that applications tend to repeatedly query non-existant paths. However, with an ever-growing dentry tree, this will also lead to unbounded memory growth due to map size growth. We had a similar issue with goferfs earlier and we had to implement an LRU cache for negative entries: https://github.com/google/gvisor/commit/95ef37391962b608cdfca9d95d32be56095e0e12.

Reply:
Good idea! Will add it to the backlog. Thanks!